src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs

```diff
@@ -111,74 +154,164 @@ internal abstract class AbstractAddParameterCodeFixProvider<
+            }
 
-                var parameters = constructor.Parameters.Select(p => p.ToDisplayString(SimpleFormat));
-                var signature = $"{type.Name}({string.Join(", ", parameters)})";
+            return null;
+        }
 
-                var title = string.Format(FeaturesResources.Add_parameter_to_0, signature);
+        private TArgumentSyntax GetArgumentInsertPositionForMethodCandidate(SemanticModel semanticModel, ISyntaxFactsService syntaxFacts, IMethodSymbol method, SeparatedSyntaxList<TArgumentSyntax> arguments, TArgumentSyntax argumentOpt)
+        {
+            var comparer = syntaxFacts.StringComparer;
+            var isNamedArgument = !string.IsNullOrWhiteSpace(syntaxFacts.GetNameForArgument(argumentOpt));
 
-                context.RegisterCodeFix(
-                    new MyCodeAction(title, c => FixAsync(document, constructor, argumentToAdd, arguments, c)),
-                    context.Diagnostics);
+            if (isNamedArgument || NonParamsParameterCount(method) < arguments.Count)
+            {
+                var argumentToAdd = DetermineFirstArgumentToAdd(
+                    semanticModel, syntaxFacts, comparer, method,
+                    arguments, argumentOpt);
+
+                if (argumentToAdd != null)
+                {
+                    if (argumentOpt != null && argumentToAdd != argumentOpt)
+                    {
+                        // We were trying to fix a specific argument, but the argument we want
+                        // to fix is something different.  That means there was an error earlier
+                        // than this argument.  Which means we're looking at a non-viable 
+                        // constructor.  Skip this one.
+                        return null;
+                    }
+                }
+
+                return argumentToAdd;
             }
+
+            return null;
         }
 
         private int NonParamsParameterCount(IMethodSymbol method)
             => method.IsParams() ? method.Parameters.Length - 1 : method.Parameters.Length;
 
-        private async Task<Document> FixAsync(
+        private async Task<Solution> FixAsync(
             Document invocationDocument,
             IMethodSymbol method,
             TArgumentSyntax argument,
             SeparatedSyntaxList<TArgumentSyntax> argumentList,
+            TTypeSyntax newTypeArgument,
             CancellationToken cancellationToken)
         {
-            var methodDeclaration = await method.DeclaringSyntaxReferences[0].GetSyntaxAsync(cancellationToken).ConfigureAwait(false);
+            var newSolution = invocationDocument.Project.Solution;
+            
+            //method.DeclaringSyntaxReferences does no longer return the two locations for partial methods so this is no longer true:
+            //https://stackoverflow.com/a/38294041
+            //Therefore we need this ugly hack:
+            var methodSymbolReferences =
+                Enumerable.Repeat(method, 1)
```

Comment:
you could just do something like "ImmutableArray.Create(method, method.PartialDefinitionPart, method.PartialImplementationPart).WhereNotNull().Distinct().ToImmutable()".

This usage of .Union and Enumerable.repeat is very unlike any of our other code.